Remove return outside of function as it is not es6 module compliant

FranzPoize opened the pull request:
Since which-collection has been added to is-equal there's a lot of module dependant on is-map. According to https://github.com/babel/babel/issues/1298 returns outside of function are not es6 module compliant. So I submit this PR to avoid the return.

I changed the if to avoid the negative condition. Please give me some feedback if you're unhappy with anything. Thank you.

ljharb commented:
This file isn’t an es6 module, it’s a CommonJS module, which it is compliant with. You shouldn’t be transpiling code you didn’t write, which includes everything in node_modules, and this package.

pocka commented:
> This file isn’t an es6 module, it’s a CommonJS module, which it is compliant with. You shouldn’t be transpiling code you didn’t write, which includes everything in node_modules, and this package.

I agree with we should not transpile dependencies. If you need to do it (ex. creating a bundle for ES5 environment), you should use options like `allowReturnOutsideFunction` instead.

But let me correct a few. CJS spec does not mention to top-level return. It works because "Node.js runs Node.js modules (≒ CJS modules) with [wrapper code](https://github.com/nodejs/node/blob/357a99293e0132ed96afde56df5ce5aea2b550cd/lib/internal/modules/cjs/loader.js#L184)". And, IIRC, both ES5 and ES6 language spec don't allow top-level return statements regardless of Scripts or Modules.  For reference, [this issue](https://github.com/mishoo/UglifyJS2/issues/288) might help (too old but I think it makes sense).

ljharb commented:
The legacy CJS spec is irrelevant; “what node supports” is what node modules require, and top level return is part of that.

node modules are not a Script, they’re a function scope - and as such, top level return is already valid. Any tool that claims to support node modules but doesn’t support top level return is lying or broken.

pocka commented:
@ljharb 
Yeah, my point was that some people (especially run JS on browser with bundlers) consider CJS as "legacy CJS spec", so it's safer to say "node modules" or something. Sorry for confusing writing :cold_sweat:

ljharb commented:
The only module types that matter (or have for years) are node modules (commonly called CJS) or native language modules (called ESM).

ljharb committed:
[Refactor] avoid top-level return, because babel and webpack are broken

Closes #5. Closes #4. Closes #3.

Closes https://github.com/inspect-js/node-deep-equal/issues/78.
Closes https://github.com/es-shims/Promise.allSettled/issues/7.
Closes https://github.com/airbnb/js-shims/issues/12.

Relates to https://github.com/inspect-js/is-set/pull/4, https://github.com/inspect-js/is-set/pull/5

Addresses https://github.com/storybookjs/storybook/issues/9154.

FranzPoize commented:
This was an enjoyable experience thank you.

ljharb commented:
Turns out that this is caused by multiple kinds of breakage:
 1. webpack, prior to v4.31, does not properly parse CJS modules
 1. babel does not properly parse CJS modules without the `allowTopLevelReturn` option enabled
 1. CRA unsafely transpiles node_modules, and does not enable the `allowTopLevelReturn` babel option
